Spurious access error typechecking java sources with protected inner class

## Reproduction steps

Scala version: 2.13.10

```java
package compiletest.a;
public class OuterParent {
    protected static class InnerParent {
    }
}
```

```java
package compiletest.b;
import compiletest.a.OuterParent;

public class OuterChild extends OuterParent {
    public static class InnerChild extends InnerParent {
    }
}
```

```scala
package client

object Client {
  new compiletest.b.OuterChild.InnerChild
}
```

## Problem

When the files are compiled jointly, scalac emits a spurious error:

```
➜ scalac --scala-version 2.13.10 -d /tmp -cp /tmp test/files/pos/java-protected-inner-static-class-ypicklejava/*
test/files/pos/java-protected-inner-static-class-ypicklejava/OuterChild.java:5: error: class InnerParent in class OuterParent cannot be accessed as a member of object compiletest.b.OuterChild from class OuterChild in package b
 Access to protected class InnerParent not permitted because
 enclosing class OuterChild in package b is not a subclass of
 class OuterParent in package a where target is defined
    public static class InnerChild extends InnerParent {
                                           ^
1 error
```

Workaround is to compile separately:

```
➜ javac -d /tmp test/files/pos/java-protected-inner-static-class-ypicklejava/*.java
➜ scalac --scala-version 2.13.10 -d /tmp -cp /tmp test/files/pos/java-protected-inner-static-class-ypicklejava/*.scala
```

--- a/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
@@ -896,7 +896,9 @@ trait Contexts { self: Analyzer =>
     private def isSubClassOrCompanion(sub: Symbol, base: Symbol) =
       sub.isNonBottomSubClass(base) ||
         (sub.isModuleClass && sub.linkedClassOfClass.isNonBottomSubClass(base)) ||
-        (base.isJavaDefined && base.isModuleClass && sub.isNonBottomSubClass(base.linkedClassOfClass))
+        (base.isJavaDefined && base.isModuleClass && (
+          sub.isNonBottomSubClass(base.linkedClassOfClass) ||
+            sub.isModuleClass && sub.linkedClassOfClass.isNonBottomSubClass(base.linkedClassOfClass)))
 
     /** Return the closest enclosing context that defines a subclass of `clazz`
      *  or a companion object thereof, or `NoContext` if no such context exists.

--- a/test/files/pos/t12673/A.java
+++ b/test/files/pos/t12673/A.java
@@ -1,0 +1,4 @@
+package compiletest.a;
+public class A {
+  protected static class InnerParent { }
+}

--- a/test/files/pos/t12673/B.java
+++ b/test/files/pos/t12673/B.java
@@ -1,0 +1,6 @@
+package compiletest.b;
+import compiletest.a.A;
+
+public class B extends A {
+  public static class InnerChild extends InnerParent { }
+}

--- a/test/files/pos/t12673/Test.scala
+++ b/test/files/pos/t12673/Test.scala
@@ -1,0 +1,5 @@
+package client
+
+object Client {
+  new compiletest.b.B.InnerChild
+}

